CRM/Event/Form/Participant.php

```diff
@@ -779,16 +779,21 @@ public static function formRule($values, $files, $self) {
       $event->id = $eventId;
       $event->find(TRUE);
 
-      if (!$event->allow_same_participant_emails && !empty($contactId) && !empty($eventId)) {
-        $cancelledStatusID = CRM_Core_PseudoConstant::getKey('CRM_Event_BAO_Participant', 'status_id', 'Cancelled');
-        $dupeCheck = new CRM_Event_BAO_Participant();
-        $dupeCheck->contact_id = $contactId;
-        $dupeCheck->event_id = $eventId;
-        $dupeCheck->whereAdd("status_id != {$cancelledStatusID} ");
-        $dupeCheck->find(TRUE);
-        if (!empty($dupeCheck->id)) {
-          $errorMsg['event_id'] = ts('This contact has already been assigned to this event.');
-        }
+      if (
+        !$event->allow_same_participant_emails
+        && !empty($contactId)
+        && !empty($eventId)
+        && CRM_Event_BAO_Participant::exists(
+          $contactId,
+          $eventId,
+          FALSE,
+          TRUE,
```

Comment:
As per discussion with @colemanw and @JoeMurray this should be `FALSE` for not taking `is_test` participants into account.